common/lib/xmodule/xmodule/modulestore/__init__.py

```diff
@@ -409,6 +410,36 @@ class ModuleStoreAssetWriteInterface(ModuleStoreAssetInterface):
     """
     The write operations for assets and asset metadata
     """
+    def _save_assets_by_type(self, course_key, asset_metadata_list, course_assets, user_id, import_only):
+        """
+        Common private method that saves/updates asset metadata items in the internal modulestore
+        structure used to store asset metadata items.
+        """
+        # Lazily create a sorted list if not already created.
+        assets_by_type = defaultdict(lambda: SortedListWithKey(course_assets.get(asset_type, []), key=itemgetter('filename')))
+
+        for asset_md in asset_metadata_list:
+            if asset_md.asset_id.course_key != course_key:
+                log.warning("Asset's course {} does not match other assets for course {} - not saved.".format(
+                    asset_md.asset_id.course_key, course_key
+                ))
+                continue
+            asset_type = asset_md.asset_id.asset_type
+            all_assets = assets_by_type[asset_type]
+            asset_idx = self._find_asset_in_list(all_assets, asset_md.asset_id)
```

Comment:
I'm not sure if this is the right place to address this, but design wise, it seems like this method (`_find_asset_in_list`) might have been better if placed onto a subclass of `SortedListWithKey` that knew how to deal with assets, rather than living on the modulestore.